### PR TITLE
User ISO Request Handler

### PR DESCRIPTION
**Pull request by thomasonw**
Added capability for user to define ISO Request-Message handler.

### CODE REVIEW

File: NMEA2000.cpp [hunk 1]
```diff
@@ -534,7 +537,13 @@ void tNMEA2000::HandleISORequest(const tN2kMsg &N2kMsg) {
       case 126998L: /* Configuration information */
         SendConfigurationInformation(iDev);
         break;
-       // ToDo: Write default handler so that user can add their own.
+      default:
+         if (RqstHandler!=0 ) RqstHandler(N2kMsg);     /* See if user wants to handle it */
```
Comment (sarfata) [inline NMEA2000.cpp]: Maybe the request handler could return true/false to indicate whether they processed the message.

If the `RqstHandler` returns `true`, you do nothing more (like you are doing now). If it returns `false`, you could then send a standard ACK.

File: NMEA2000.cpp [hunk 2]
```diff
@@ -707,6 +721,13 @@ void SetN2kPGN126996(tN2kMsg &N2kMsg, unsigned int N2kVersion, unsigned int Prod
 
 //*****************************************************************************
 // Iso request
+void SetN2kPGN59904(tN2kMsg &N2kMsg, uint8_t Destination, unsigned long RequestedPGN) {
```
Comment (sarfata) [inline NMEA2000.cpp]: This seems unrelated to this PR. Did you mean to include this?

File: NMEA2000.h [hunk 3]
```diff
@@ -345,6 +349,13 @@ inline void SetN2kProductInformation(tN2kMsg &N2kMsg, unsigned int N2kVersion, u
                   SertificationLevel,LoadEquivalency);
 }
 
+
+void SetN2kPGN59904(tN2kMsg &N2kMsg, uint8_t Destination, unsigned long RequestedPGN);
+
+inline void SetN2kPGNISORequest(tN2kMsg &N2kMsg, uint8_t Destination, unsigned long RequestedPGN) {
```
Comment (sarfata) [inline NMEA2000.h]: This seems unrelated to this PR. Did you mean to include this?

File: NMEA2000.cpp (hunk 1 above)
Comment (thomasonw) [inline NMEA2000.cpp]: That is an idea, would save us from replicating the NAK code in our handler.
Perhaps a downside is the RqstHandler and the _MsgHandler functions would then be asymmetrical (One BOOL and the other VOID )

File: NMEA2000.cpp (hunk 2 above)
Comment (thomasonw) [inline NMEA2000.cpp]: Yes, these complete the Request API:  Able to send and receive.

File: NMEA2000.h (hunk 3 above)
Comment (thomasonw) [inline NMEA2000.h]: Yes, these complete the Request API:  Able to send and receive.

Comment (ttlappalainen) [inline NMEA2000.h]: Why we would't define
    // ISO Request handler. Handler must return true, if ith has handled request completely
    // Handler should respond to Destination
    bool(*ISORequestHandler)(unsigned long RequestedPGN, unsigned char Destination, int DeviceIndex);

   Then handler does not need to parse request msg again.
   DeviceIndex is the id for multidevice system, which is still under construction.

Comment (thomasonw) [inline NMEA2000.h]: I like that!   I might use 'Requester' instead of 'Destination' for the 2nd parameter.   Destination sounds to me like our address, not the address of the node making the request.  So:

ool(*ISORequestHandler)(unsigned long RequestedPGN, unsigned char Requester, int DeviceIndex);

A detail:  If the NAK is going to be generated in the NMEA2000.ccp module, It is my understanding the NAK is only sent when the  request was directed to a specific node.  However, for Broadcast Requests (destination address = 0xFF), no such NAK should be sent.

Comment (ttlappalainen) [inline NMEA2000.h]: Good point, Requester is better.

I think you are right about sending NAK only for non broadcasts. But can not say for sure, since I do not have manuals. There would be no sense for responding NAK to broadcast, but not all on N2k has sense anyway.

Comment (thomasonw) [inline NMEA2000.h]: Here is a 'back-door' resource I am using for J1939:  the RV-C spec  ([http://www.rv-c.com/?q=node/75](http://www.rv-c.com/?q=node/75))    Like NMEA2000, it is built on top of J1939,and as such it documents several of the underlying foundation messages, like Requests and addresses, etc.

ref: 3.2.4.3 #1 concerning how to respond of a Request.

Comment (ttlappalainen) [inline NMEA2000.h]: Thanks. I forgot that I can use J1939 documents even I have already used them e.g. for address claiming. Getting old...

**Comment by ttlappalainen** [comment]
Hi,

Any progress with this?
- bool(*ISORequestHandler)(unsigned long RequestedPGN, unsigned char Requester, int DeviceIndex);
- NAK sending to library, if handler returns false.

**Comment by thomasonw** [comment]
Hello – sorry for the delay, we are currently in an area with poor internet connection.   POP3 Email works at times (with several day gaps between windows of opportunities), and web based portals, like GitHub, are out of the question.

 

Late next week should have more reliable connection and I will finish this up then.

 

-al-

**Comment by thomasonw** [comment]
Closing this pull request.  New one submitted with these comments and suggestions.
